zap/zap.go

```diff
@@ -14,89 +15,173 @@ var (
+	return zpl.WithLevel(slog.Debug)
 }
 
 // Info returns a new logger set to add entries as level Info
-func (zl *Logger) Info() slog.Logger {
-	return zl.WithLevel(slog.Info)
+func (zpl *Logger) Info() slog.Logger {
+	return zpl.WithLevel(slog.Info)
 }
 
 // Warn returns a new logger set to add entries as level Warn
-func (zl *Logger) Warn() slog.Logger {
-	return zl.WithLevel(slog.Warn)
+func (zpl *Logger) Warn() slog.Logger {
+	return zpl.WithLevel(slog.Warn)
 }
 
 // Error returns a new logger set to add entries as level Error
-func (zl *Logger) Error() slog.Logger {
-	return zl.WithLevel(slog.Error)
+func (zpl *Logger) Error() slog.Logger {
+	return zpl.WithLevel(slog.Error)
 }
 
 // Fatal returns a new logger set to add entries as level Fatal
-func (zl *Logger) Fatal() slog.Logger {
-	return zl.WithLevel(slog.Fatal)
+func (zpl *Logger) Fatal() slog.Logger {
+	return zpl.WithLevel(slog.Fatal)
 }
 
 // WithLevel returns a new logger set to add entries to the specified level
-func (zl *Logger) WithLevel(level slog.LogLevel) slog.Logger {
-	return zl
+func (zpl *Logger) WithLevel(level slog.LogLevel) slog.Logger {
+
+	if zpl.cfg != nil {
+
+		var levels = []zapcore.Level{
+			slog.UndefinedLevel: zapcore.InvalidLevel,
+			slog.Fatal:          zapcore.FatalLevel,
+			slog.Error:          zapcore.ErrorLevel,
+			slog.Warn:           zapcore.WarnLevel,
+			slog.Info:           zapcore.InfoLevel,
+			slog.Debug:          zapcore.DebugLevel,
+		}
+
+		if level < -1 || int(level) >= len(levels) {
+			// fix your code
+			err := fmt.Errorf("slog: invalid log level %v", level)
+			zpl.logger.Panic(err.Error())
+
+		} else if zpl.logger.Core().Enabled(levels[level]) {
+			zpllevel := levels[level]
+			// we set cfg level here
+			zpl.cfg.Level.SetLevel(zpllevel)
+		}
+	}
+
+	return zpl
 }
 
 // WithStack attaches a call stack to a new logger
-func (zl *Logger) WithStack(skip int) slog.Logger {
-	return zl
+func (zpl *Logger) WithStack(skip int) slog.Logger {
+	zpl.logger = zpl.logger.WithOptions(zap.AddStacktrace(zapcore.InfoLevel))
+	return zpl
 }
 
 // WithField returns a new logger with a field attached
-func (zl *Logger) WithField(label string, value any) slog.Logger {
-	return zl
+func (zpl *Logger) WithField(label string, value any) slog.Logger {
+	zpl.logger = zpl.logger.With(zap.Any(label, value))
+	return zpl
 }
 
 // WithFields returns a new logger with a set of fields attached
-func (zl *Logger) WithFields(fields map[string]any) slog.Logger {
-	return zl
+func (zpl *Logger) WithFields(fields map[string]any) slog.Logger {
+	zs := make([]zap.Field, 0)
+	for k, v := range fields {
+		zs = append(zs, zap.Any(k, v))
+	}
+	zpl.logger = zpl.logger.With(zs...)
+	return zpl
+}
+
+// New creates a slog.Logger adaptor using a zap as backend. If
+// none was passed it will create an opiniated new one, and you will
+// have no option to change its config
+func New(logger *zap.Logger) slog.Logger {
+
+	return newLogger(logger, "info")
```

Comment:
wouldn't it make more sense to let the user choose the level? maybe using our constants?